server/models/service-process.js

```diff
@@ -0,0 +1,132 @@
+var async = require('async');
+var debug = require('debug')('strong-mesh-models:server:service-process');
+
+module.exports = function(ServiceProcess) {
+  function recordFork(instanceId, pInfo, callback) {
+    debug('Process forked: worker id %d pid %d ppid %d',
+      pInfo.id, pInfo.pid, pInfo.ppid);
+
+    var proc = new ServiceProcess({
+      pid: pInfo.pid,
+      parentPid: pInfo.ppid,
+      workerId: pInfo.id,
+      serviceInstanceId: instanceId,
+      startTime: new Date()
+    });
+
+    proc.save(function(err, proc) {
+      debug('upsert Process: %j', err || proc);
+      if (err) return callback(err);
+      callback();
+    });
+  }
+  ServiceProcess.recordFork = recordFork;
+
+  function recordExit(instanceId, pInfo, callback) {
+    debug('Process exited: worker id %d pid %d reason %s suicide? %s',
+      pInfo.id, pInfo.pid, pInfo.reason, pInfo.suicide);
+
+    async.waterfall([
+      _findProcess(instanceId, +pInfo.id, pInfo.pid),
+      updateProcess,
+      updateChildren
+    ], function ensureSave(err, proc) {
+      debug('on exit of %j, save Process: %j', pInfo, err || proc);
+      callback(err);
+    });
+
+    function updateProcess(proc, asyncCb) {
+      if (!proc) {
+        // Ignore process if it is not in the DB. Defensive coding against race
+        // conditions, shouldn't happen, never seen, but lets handle it.
+        debug('exit event for an unknown process: %j', pInfo);
+        return asyncCb();
```

Comment:
Doesn't this break `updateChildren()` since it results in its `proc` argument being `undefined`, which it doesn't guard against?